File: results/2026-04-02_163146/tasks/12-pr-label-assigner/csharp-script-sonnet/generated-code/PrLabelAssigner.Tests/LabelAssigner.cs
// LabelAssigner: the core engine that assigns labels to a set of changed files.
//
// Algorithm:
//   1. Sort rules by priority (ascending; lower number = evaluated first).
//   2. For each file path, test every rule in priority order.
//   3. When a rule's glob pattern matches the file path, add the rule's label to the result set.
//   4. Deduplicate labels (a HashSet handles this automatically).
//   5. Return the final label set.
//
// Design decisions:
//   - Multiple labels CAN be applied to a single PR (union of all matching rule labels).
//   - Priority controls evaluation order but does NOT suppress other labels.
//     All matching rules contribute their labels regardless of priority.
//   - Labels are case-sensitive and returned as a sorted, read-only set.

namespace PrLabelAssigner;

public class LabelAssigner
{
    // Rules sorted by priority (lower number = higher priority = evaluated first)
    private readonly IReadOnlyList<LabelRule> _rules;

    public LabelAssigner(IEnumerable<LabelRule> rules)
    {
        _rules = rules
            .OrderBy(r => r.Priority)
            .ThenBy(r => r.Pattern) // stable secondary sort for determinism
            .ToList()
            .AsReadOnly();
    }

    /// <summary>
    /// Assigns labels to a PR based on its changed file paths.
    /// </summary>
    /// <param name="filePaths">The list of file paths changed in the PR.</param>
    /// <returns>A read-only, deduplicated set of labels to apply.</returns>
    public IReadOnlySet<string> AssignLabels(IEnumerable<string> filePaths)
    {
        var labels = new SortedSet<string>(StringComparer.Ordinal);

        foreach (var filePath in filePaths)
        {
            foreach (var rule in _rules)
            {
                if (GlobMatcher.IsMatch(filePath, rule.Pattern))
                {
                    labels.Add(rule.Label);
                }
            }
        }

        return labels;
    }
}

File: results/2026-04-17_004319/tasks/12-pr-label-assigner/default-sonnet46-1m-medium/generated-code/label_assigner.py
"""
PR Label Assigner — assigns labels to a pull request based on which files changed.

Rules map glob patterns to labels with a priority number (lower = higher priority).
Multiple rules can match; all their labels are collected and deduplicated.
Output is ordered by the priority of the first rule that produced each label.
"""

import argparse
import fnmatch
import json
import sys


def match_file(file_path: str, pattern: str) -> bool:
    """Return True if file_path matches the glob pattern.

    Uses fnmatch which lets * match path separators, so docs/** correctly
    matches docs/sub/dir/file.md without needing pathlib.
    """
    return fnmatch.fnmatch(file_path, pattern)


def load_rules(rules_config: list[dict]) -> list[dict]:
    """Return rules sorted by priority ascending (1 = highest priority).

    Rules without an explicit priority field are placed last (fallback 999).
    The original list is not modified.
    """
    return sorted(rules_config, key=lambda r: r.get("priority", 999))


def assign_labels(files: list[str], rules: list[dict]) -> list[str]:
    """Assign labels to a PR given its changed files and a set of rules.

    Algorithm:
      1. Sort rules by priority (lowest number first).
      2. For each rule, check whether ANY file matches its pattern.
      3. If so, add the rule's label to the output (skip if already present).
      4. Return the deduplicated label list in priority order.
    """
    sorted_rules = load_rules(rules)
    seen: set[str] = set()
    labels: list[str] = []

    for rule in sorted_rules:
        pattern = rule["pattern"]
        label = rule["label"]
        for file_path in files:
            if match_file(file_path, pattern):
                if label not in seen:
                    seen.add(label)
                    labels.append(label)
                break  # one matching file is enough to activate this rule

    return labels


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Assign PR labels based on changed file paths."
    )
    parser.add_argument("input_file", help="JSON file with 'files' and 'rules' keys")
    parser.add_argument(
        "--format",
        choices=["list", "csv", "json"],
        default="list",
        help="Output format (default: list)",
    )
    args = parser.parse_args()

    try:
        with open(args.input_file) as fh:
            data = json.load(fh)
    except FileNotFoundError:
        print(f"Error: input file '{args.input_file}' not found", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as exc:
        print(f"Error: invalid JSON in '{args.input_file}': {exc}", file=sys.stderr)
        sys.exit(1)

    files = data.get("files", [])
    rules = data.get("rules", [])

    if not isinstance(files, list):
        print("Error: 'files' must be a JSON array", file=sys.stderr)
        sys.exit(1)
    if not isinstance(rules, list):
        print("Error: 'rules' must be a JSON array", file=sys.stderr)
        sys.exit(1)

    result = assign_labels(files, rules)

    if args.format == "csv":
        print(",".join(result))
    elif args.format == "json":
        print(json.dumps(result))
    else:
        for label in result:
            print(label)

    # Always emit a machine-parseable summary line for the act test harness
    print(f"ASSIGNED_LABELS: {','.join(result)}")


if __name__ == "__main__":
    main()
